docs/source/tutorial/basic.rst

```diff
@@ -2,21 +2,586 @@
+
+The command will write the provenance graph to a ``.pdf`` file.
+Use your favorite PDF viewer to have a look.
+It should look something like the graph shown in :numref:`fig_calcfun_graph`.
+
+.. _fig_calcfun_graph:
+.. figure:: include/calcfun_graph.png
+    :scale: 50
+    :align: center
+
+    Provenance graph of the ``multiply`` calculation function.
+
+.. note:: Remember that the PK of the ``CalcJob`` is most likely different for your database.
+
+CalcJobs
+========
+
+When running calculations that require an external code or run on a remote machine, a simple calculation function is no longer sufficient.
+For this purpose, AiiDA provides the ``CalcJob`` process class.
+
+To run a ``CalcJob``, you need to set up two things: a ``computer`` for the calculation to run on, and the ``code`` that you want the ``CalcJob`` to run.
+If you're running this tutorial in the Quantum Mobile VM or on Binder, these have been pre-configured for you. If you're running on your own machine, you can follow the instructions in the panel below:
+
+.. accordion:: Install localhost computer and code
+
+    Let's begin by setting up the computer using the ``verdi computer`` subcommand:
+
+    .. code-block:: console
+
+        $ verdi computer setup -L tutor -H localhost -T local -S direct -w `echo $PWD/work` -n
+        $ verdi computer configure local tutor --safe-interval 5 -n
+
+    The first commands sets up the computer with the following options:
+
+    * *label* (``-L``): tutor
+    * *hostname* (``-H``): localhost
+    * *transport* (``-T``): local
+    * *scheduler* (``-S``): direct
+    * *work-dir* (``-w``): The ``work`` subdirectory of the current directory
+
+    The second command *configures* the computer with a minimum interval between connections (``--safe-interval``) of 5 seconds.
+    For both commands, the *non-interactive* option (``-n``) is added to not prompt for extra input.
+
+    Next, let's set up the code we're going to use for the tutorial:
+
+    .. code-block:: console
+
+        $ verdi code setup -L add --on-computer --computer=tutor -P arithmetic.add --remote-abs-path=/bin/bash -n
+
+    This command sets up a code with *label* ``add`` on the *computer* ``tutor``, using the *plugin* ``arithmetic.add``.
+
+.. note::
+    A typical real-world example of a computer is a remote supercomputing facility.
+    Codes can be anything from a Python script to powerful *ab initio* codes such as Quantum Espresso or machine learning tools like Tensorflow.
+
+Let's have a look at the codes that are available to us:
+
+.. code:: console
+
+    $ verdi code list
+
+You can see a single code ``add@tutor``, with PK = 5, in the printed list.
+This code allows us to add two integers together.
+The ``add@tutor`` identifier indicates that the code with label ``add`` is run on the computer with label ``tutor``.
+To see more details about the computer, you can use the following ``verdi`` command:
+
+.. code:: console
+
+    $ verdi computer show tutor
+
```

Comment:
Alright, I've added the output. It's probably a good idea to also explain the fact that the `tutor` computer has PK = 1, same as the ``Int`` node that they made at the start of the tutorial. I know that nodes, computers and groups are different types of entities in the database. Is that why they can have the same PK? How would I best explain this in the tutorial?

Reply:
I actually didn't know that and cannot provide any guidance on this. Maybe @sphuber can help? So far we've been promoting the message that PKs are unique within the profile space.

Reply:
Yes, the PKs are unique for each profile for each "entity" as long as they are stored in a separate database table, simply because they are the unique identifiers in said table. The following entities have their own table, AuthInfo, Comment, Computer, Group, Node, Link, Log, Node, Setting and User, and so they have their own unique set of pks. Note that sublcasses of each entity are stored in the same table.

Of course we should not explain it in this way or with this amount of details to the user. The most relevant entities for a user would be Computer, Group and Node. So maybe we can add a quick note where we say that PKs are unique to each database, that they are even unique within a database for different entities.

Reply:
For those users familiar with databases, it will make immediate sense that the primary key (PK) is unique for a given table and that computers and nodes, etc. are stored in different tables and can therefore have overlapping PKs.

So let's make sure that we mention this such that "expert" users can contextualize this and for all other users we will just say that PKs are unique to each profile for each "class".

Reply:
I was thinking of adding a note after they see the output of `verdi computer show tutor`, as they will probably wonder about the PK at that point. Something along the lines of:

```
.. note:: 

    You may have noticed that the PK of the ``tutor`` computer is 1, same as the ``Int`` node we created at the start of this tutorial.
    This is because different entities, such as nodes, computers and groups, are stored in different tables of the database.
    So, the PKs for each entity type are unique for each database, but entities of different types can have the same PK within one database.
```